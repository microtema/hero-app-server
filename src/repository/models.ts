import readdirSync from 'recursive-readdir-synchronous';
import Model, {Sequelize} from 'sequelize';

const {DATABASE, DATABASE_USER, DATABASE_PWD} = process.env;

const sequelize = new Sequelize(DATABASE, DATABASE_USER, DATABASE_PWD,
    {
        dialect: 'postgres',
    },
);

console.log('Connected to database successfully');

const filterModel = (it) => it.match('.orm.ts');
const cleanupModelPath = (it) => '..' + it.substring(it.indexOf('/src') + 4).replace('.ts', '');
const importModel = (it) => sequelize.import(it);
const capitalizeFirstLetter = (str) => str.charAt(0).toUpperCase() + str.slice(1);

// Note: Autowired models
const models = readdirSync('./src')
    .filter(filterModel)
    .map(cleanupModelPath)
    .map(importModel)
    .reduce((map, it) => {
        const key = capitalizeFirstLetter(it.name + 'Model');
        map[key] = it;
        return map;
    }, {});

const associate = (modelName) => {
    models[modelName].associate(models);
};

Object.keys(models).forEach(associate);

export {sequelize};

export default models;
