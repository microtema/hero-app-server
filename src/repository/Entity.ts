export interface Entity<ID> {
    id: ID;
}
