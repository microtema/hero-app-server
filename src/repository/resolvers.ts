import {merge} from 'lodash';
import {loadModules} from '../util/File.util';

export default loadModules('.resolver.ts').reduce((map, it) => merge(map, it), {});
