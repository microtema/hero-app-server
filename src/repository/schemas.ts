import {gql} from 'apollo-server-express';
import {loadModules} from '../util/File.util';

const linkSchema = gql`
    type Query {
        _: Boolean
    }
    type Mutation {
        _: Boolean
    }
    type Subscription {
        _: Boolean
    }
`;

export default [linkSchema, ...loadModules('.schema.ts')];
