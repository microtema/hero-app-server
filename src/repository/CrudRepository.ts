import {DestroyOptions, FindOptions, Identifier, UpdateOptions} from 'sequelize';
import {Entity, EntityStatic} from './Entity';

abstract class CrudRepository<T extends Entity<ID>, ID extends Identifier> {

    /**
     * Find all Authors filtered by name
     *
     * @param options may be undefined or null
     */
    public findAll(options?: any) {

        return this.model().findAll(options || {});
    }

    /**
     * Find entity by id
     *
     * @param id may not be null
     * @param options may be null
     */
    public findByPk(id: ID, options?: any) {

        options = options || {
            where: {
                id,
            },
        } as FindOptions;

        return this.model().findByPk(id, options);
    }

    /**
     * Save new entity
     *
     * @param entity may not be null
     */
    public save(entity: T) {

        return this.model().create(entity);
    }

    /**
     * Update new Author
     *
     * @param entity may not be null
     */
    public update(entity: T) {

        const predicate = {
            where: {
                id: entity.id,
            },
        } as UpdateOptions;

        return this.model().update(entity, predicate);
    }

    /**
     * Delete enity by Id
     *
     * @param id may not be null
     */
    public delete(id: ID) {

        return this.model().destroy({where: {id}} as DestroyOptions);
    }

    public abstract model(): EntityStatic<ID>;
}

export default CrudRepository;
