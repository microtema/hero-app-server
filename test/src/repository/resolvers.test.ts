import {default as sut} from '../../../src/repository/resolvers';

describe('Resolvers', () => {

    it('Should be defined', () => {
        expect(sut).toBeDefined();
    });

    it('Should match', () => {
        expect(sut).toMatchSnapshot();
    });
});
