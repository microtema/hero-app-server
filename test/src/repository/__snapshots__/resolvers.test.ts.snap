// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resolvers Should match 1`] = `
Object {
  "Mutation": Object {
    "createAddress": [Function],
    "createAuthor": [Function],
    "createBook": [Function],
    "deleteAddress": [Function],
    "deleteAuthor": [Function],
    "deleteBook": [Function],
    "updateAddress": [Function],
    "updateAuthor": [Function],
    "updateBook": [Function],
  },
  "Query": Object {
    "address": [Function],
    "addresses": [Function],
    "author": [Function],
    "authors": [Function],
    "book": [Function],
    "books": [Function],
  },
}
`;
